web/src/client/storage.js

```diff
@@ -50,10 +50,10 @@ class StorageBaseClient {
    */
   async getProposal() {
     const storageProxy = await this.client.proxy(PROPOSAL_CALCULATOR_IFACE, STORAGE_PATH);
-    const proposalProxy = await this.client.proxy(PROPOSAL_IFACE);
+    const proposalProxy = await this.client.proxy(PROPOSAL_IFACE).catch(() => undefined);
 
     // Check whether proposal object is already exported
-    if (!proposalProxy.valid) return {};
+    if (!proposalProxy?.valid) return {};
```

Comment:
Looks like we cannot comment in [untouched lines yet](https://github.com/github/roadmap/issues/347), so let's use this one for commenting another approach we could follow in this method.

The idea is simple but can be split in two

* To use a try/catch block for surrounding the full method. I.e., returning an _empty proposal_ whenever an error happens. Note that here we are getting a broken promise because `cockpit.dbus.proxy` is crashing due to something (?) missing in the Dbus side.

  <details>
  <summary>Click to show/hide a diff</summary>

  ---

  ```diff
  diff --git a/web/src/client/storage.js b/web/src/client/storage.js
  index 086e68a38..dccf0e08c 100644
  --- a/web/src/client/storage.js
  +++ b/web/src/client/storage.js
  @@ -49,51 +49,57 @@ class StorageBaseClient {
      * @return {Promise<object>}
      */
     async getProposal() {
  -    const storageProxy = await this.client.proxy(PROPOSAL_CALCULATOR_IFACE, STORAGE_PATH);
  -    const proposalProxy = await this.client.proxy(PROPOSAL_IFACE).catch(() => undefined);
  -
  -    // Check whether proposal object is already exported
  -    if (!proposalProxy?.valid) return {};
  -
  -    const volume = dbusVolume => {
  -      const valueFrom = dbusValue => dbusValue?.v;
  +    const emptyProposal = {};
  +
  +    try {
  +      const storageProxy = await this.client.proxy(PROPOSAL_CALCULATOR_IFACE, STORAGE_PATH);
  +      const proposalProxy = await this.client.proxy(PROPOSAL_IFACE);
  +
  +      // Check whether proposal object is already exported
  +      if (!proposalProxy.valid) return emptyProposal;
  +
  +      const volume = dbusVolume => {
  +        const valueFrom = dbusValue => dbusValue?.v;
  +
  +        const valuesFrom = (dbusValues) => {
  +          if (dbusValues === undefined) return [];
  +          return dbusValues.v.map(valueFrom);
  +        };
  +
  +        return {
  +          mountPoint: valueFrom(dbusVolume.MountPoint),
  +          optional: valueFrom(dbusVolume.Optional),
  +          deviceType: valueFrom(dbusVolume.DeviceType),
  +          encrypted: valueFrom(dbusVolume.Encrypted),
  +          fsTypes: valuesFrom(dbusVolume.FsTypes),
  +          fsType: valueFrom(dbusVolume.FsType),
  +          minSize: valueFrom(dbusVolume.MinSize),
  +          maxSize: valueFrom(dbusVolume.MaxSize),
  +          fixedSizeLimits: valueFrom(dbusVolume.FixedSizeLimits),
  +          adaptiveSizes: valueFrom(dbusVolume.AdaptiveSizes),
  +          snapshots: valueFrom(dbusVolume.Snapshots),
  +          snapshotsConfigurable: valueFrom(dbusVolume.SnapshotsConfigurable),
  +          snapshotsAffectSizes: valueFrom(dbusVolume.SnapshotsAffectSizes),
  +          sizeRelevantVolumes: valueFrom(dbusVolume.SizeRelevantVolumes)
  +        };
  +      };
   
  -      const valuesFrom = (dbusValues) => {
  -        if (dbusValues === undefined) return [];
  -        return dbusValues.v.map(valueFrom);
  +      const action = dbusAction => {
  +        const { Text: { v: textVar }, Subvol: { v: subvolVar }, Delete: { v: deleteVar } } = dbusAction;
  +        return { text: textVar, subvol: subvolVar, delete: deleteVar };
         };
   
         return {
  -        mountPoint: valueFrom(dbusVolume.MountPoint),
  -        optional: valueFrom(dbusVolume.Optional),
  -        deviceType: valueFrom(dbusVolume.DeviceType),
  -        encrypted: valueFrom(dbusVolume.Encrypted),
  -        fsTypes: valuesFrom(dbusVolume.FsTypes),
  -        fsType: valueFrom(dbusVolume.FsType),
  -        minSize: valueFrom(dbusVolume.MinSize),
  -        maxSize: valueFrom(dbusVolume.MaxSize),
  -        fixedSizeLimits: valueFrom(dbusVolume.FixedSizeLimits),
  -        adaptiveSizes: valueFrom(dbusVolume.AdaptiveSizes),
  -        snapshots: valueFrom(dbusVolume.Snapshots),
  -        snapshotsConfigurable: valueFrom(dbusVolume.SnapshotsConfigurable),
  -        snapshotsAffectSizes: valueFrom(dbusVolume.SnapshotsAffectSizes),
  -        sizeRelevantVolumes: valueFrom(dbusVolume.SizeRelevantVolumes)
  +        availableDevices: storageProxy.AvailableDevices.map(([id, label]) => ({ id, label })),
  +        candidateDevices: proposalProxy.CandidateDevices,
  +        lvm: proposalProxy.LVM,
  +        encryptionPassword: proposalProxy.EncryptionPassword,
  +        volumes: proposalProxy.Volumes.map(volume),
  +        actions: proposalProxy.Actions.map(action)
         };
  -    };
  -
  -    const action = dbusAction => {
  -      const { Text: { v: textVar }, Subvol: { v: subvolVar }, Delete: { v: deleteVar } } = dbusAction;
  -      return { text: textVar, subvol: subvolVar, delete: deleteVar };
  -    };
  -
  -    return {
  -      availableDevices: storageProxy.AvailableDevices.map(([id, label]) => ({ id, label })),
  -      candidateDevices: proposalProxy.CandidateDevices,
  -      lvm: proposalProxy.LVM,
  -      encryptionPassword: proposalProxy.EncryptionPassword,
  -      volumes: proposalProxy.Volumes.map(volume),
  -      actions: proposalProxy.Actions.map(action)
  -    };
  +    } catch {
  +      return emptyProposal;
  +    }
     }
   
     /**
   ```
  </details>  

* To move these _auxiliary_ method (`volume`, `actions`) outside. This could apply for the _dbusVolume_ and _dbusVolumes_ in the method below.


  <details>
  <summary>Click to show/hide a diff</summary>

  ---

  ```diff

    diff --git a/web/src/client/storage.js b/web/src/client/storage.js
  index 086e68a38..6086f5d46 100644
  --- a/web/src/client/storage.js
  +++ b/web/src/client/storage.js
  @@ -43,57 +43,63 @@ class StorageBaseClient {
       this.client = new DBusClient(STORAGE_SERVICE, address);
     }
   
  +  volume(dbusVolume) {
  +    const valueFrom = dbusValue => dbusValue?.v;
  +
  +    const valuesFrom = (dbusValues) => {
  +      if (dbusValues === undefined) return [];
  +      return dbusValues.v.map(valueFrom);
  +    };
  +
  +    return {
  +      mountPoint: valueFrom(dbusVolume.MountPoint),
  +      optional: valueFrom(dbusVolume.Optional),
  +      deviceType: valueFrom(dbusVolume.DeviceType),
  +      encrypted: valueFrom(dbusVolume.Encrypted),
  +      fsTypes: valuesFrom(dbusVolume.FsTypes),
  +      fsType: valueFrom(dbusVolume.FsType),
  +      minSize: valueFrom(dbusVolume.MinSize),
  +      maxSize: valueFrom(dbusVolume.MaxSize),
  +      fixedSizeLimits: valueFrom(dbusVolume.FixedSizeLimits),
  +      adaptiveSizes: valueFrom(dbusVolume.AdaptiveSizes),
  +      snapshots: valueFrom(dbusVolume.Snapshots),
  +      snapshotsConfigurable: valueFrom(dbusVolume.SnapshotsConfigurable),
  +      snapshotsAffectSizes: valueFrom(dbusVolume.SnapshotsAffectSizes),
  +      sizeRelevantVolumes: valueFrom(dbusVolume.SizeRelevantVolumes)
  +    };
  +  }
  +
  +  action(dbusAction) {
  +    const { Text: { v: textVar }, Subvol: { v: subvolVar }, Delete: { v: deleteVar } } = dbusAction;
  +    return { text: textVar, subvol: subvolVar, delete: deleteVar };
  +  }
  +
     /**
      * Returns storage proposal values
      *
      * @return {Promise<object>}
      */
     async getProposal() {
  -    const storageProxy = await this.client.proxy(PROPOSAL_CALCULATOR_IFACE, STORAGE_PATH);
  -    const proposalProxy = await this.client.proxy(PROPOSAL_IFACE).catch(() => undefined);
  +    const emptyProposal = {};
   
  -    // Check whether proposal object is already exported
  -    if (!proposalProxy?.valid) return {};
  +    try {
  +      const storageProxy = await this.client.proxy(PROPOSAL_CALCULATOR_IFACE, STORAGE_PATH);
  +      const proposalProxy = await this.client.proxy(PROPOSAL_IFACE);
   
  -    const volume = dbusVolume => {
  -      const valueFrom = dbusValue => dbusValue?.v;
  -
  -      const valuesFrom = (dbusValues) => {
  -        if (dbusValues === undefined) return [];
  -        return dbusValues.v.map(valueFrom);
  -      };
  +      // Check whether proposal object is already exported
  +      if (!proposalProxy.valid) return emptyProposal;
   
         return {
  -        mountPoint: valueFrom(dbusVolume.MountPoint),
  -        optional: valueFrom(dbusVolume.Optional),
  -        deviceType: valueFrom(dbusVolume.DeviceType),
  -        encrypted: valueFrom(dbusVolume.Encrypted),
  -        fsTypes: valuesFrom(dbusVolume.FsTypes),
  -        fsType: valueFrom(dbusVolume.FsType),
  -        minSize: valueFrom(dbusVolume.MinSize),
  -        maxSize: valueFrom(dbusVolume.MaxSize),
  -        fixedSizeLimits: valueFrom(dbusVolume.FixedSizeLimits),
  -        adaptiveSizes: valueFrom(dbusVolume.AdaptiveSizes),
  -        snapshots: valueFrom(dbusVolume.Snapshots),
  -        snapshotsConfigurable: valueFrom(dbusVolume.SnapshotsConfigurable),
  -        snapshotsAffectSizes: valueFrom(dbusVolume.SnapshotsAffectSizes),
  -        sizeRelevantVolumes: valueFrom(dbusVolume.SizeRelevantVolumes)
  +        availableDevices: storageProxy.AvailableDevices.map(([id, label]) => ({ id, label })),
  +        candidateDevices: proposalProxy.CandidateDevices,
  +        lvm: proposalProxy.LVM,
  +        encryptionPassword: proposalProxy.EncryptionPassword,
  +        volumes: proposalProxy.Volumes.map(this.volume),
  +        actions: proposalProxy.Actions.map(this.action)
         };
  -    };
  -
  -    const action = dbusAction => {
  -      const { Text: { v: textVar }, Subvol: { v: subvolVar }, Delete: { v: deleteVar } } = dbusAction;
  -      return { text: textVar, subvol: subvolVar, delete: deleteVar };
  -    };
  -
  -    return {
  -      availableDevices: storageProxy.AvailableDevices.map(([id, label]) => ({ id, label })),
  -      candidateDevices: proposalProxy.CandidateDevices,
  -      lvm: proposalProxy.LVM,
  -      encryptionPassword: proposalProxy.EncryptionPassword,
  -      volumes: proposalProxy.Volumes.map(volume),
  -      actions: proposalProxy.Actions.map(action)
  -    };
  +    } catch {
  +      return emptyProposal;
  +    }
     }
   
     /**
  ```
  </details>

I'm not saying that the method extraction should be done in that way. At the time of writing I have not clear how to do it (as it is in the diff but with proper documentation? using [private methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields)? moving them even further to an storage utils module?) I'm just comenting few changes I believe could simplify both, the code and the test.

But sadly I have no enough understanding about what is going wrong in the backend and that's why I'd like your thoughs here :)

Reply:
I would limit to fix the original problem. About moving those functions, I do not have a strong opinion. Both of them look good to me.